smtp-server/src/lib.rs

```diff
@@ -536,33 +572,31 @@ where
         match cmd {
             None => (),
 
-            // TODO: find some way to unify with the below branch
-            Some(Command::Ehlo { hostname }) => match conn_meta.hello {
-                Some(_) => {
-                    send_reply!(io, cfg.already_did_hello(&mut conn_meta)).await?;
-                }
-                None => dispatch_decision! {
-                    cfg.filter_hello(true, hostname.into_owned(), &mut conn_meta)
-                        .await,
-                    Accept(reply, res) => {
-                        conn_meta.hello = Some(res);
-                        send_reply!(io, reply).await?;
+            Some(cmd @ (Command::Ehlo { .. } | Command::Helo { .. } | Command::Lhlo { .. })) => {
+                let (cmd_proto, is_extended, hostname) = match cmd {
+                    Command::Ehlo { hostname } => (ProtocolName::Smtp, true, hostname),
+                    Command::Helo { hostname } => (ProtocolName::Smtp, false, hostname),
+                    Command::Lhlo { hostname } => (ProtocolName::Lmtp, true, hostname),
+                    _ => unreachable!(),
+                };
+                if cmd_proto == <Cfg::Protocol as Protocol<'static>>::PROTOCOL {
```

Comment:
Nit: swap the if and the else branches, so that the shorter branch is on top for better readability